File: native-views/react-native-native-list/ios/NativeListImageSlot.swift
import CryptoKit
import OneKeyImage
import UIKit

// Request lifetime is independent of row action epochs. An unchanged descriptor
// can outlive a text-only rebind; a new row/slot/source always invalidates it.
final class NativeListImageSlot {
  let view = OneKeyImageReusableView(frame: .zero)
  private var identity: String?
  private var epoch = 0
  private var retry: DispatchWorkItem?
  private var completion: ((Bool) -> Void)?
  private var result: Bool?

  static func signature(_ value: [String: Any]) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: value, options: [.sortedKeys])
    else { return "" }
    return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }

  func bind(
    _ source: [String: Any], key: String, variant: String = "generic",
    fit: String? = nil, placeholder: String = "#0000000F", completion: ((Bool) -> Void)? = nil
  ) {
    self.completion = completion
    let next = Self.signature([
      "source": source, "key": key, "variant": variant,
      "fit": fit ?? source.string("contentFit", default: "cover"), "placeholder": placeholder,
    ])
    if identity == next {
      if let result { completion?(result) }
      return
    }
    recycle()
    self.completion = completion
    identity = next
    configure(
      source, key: key, variant: variant, fit: fit, placeholder: placeholder, attempt: 0,
      token: epoch)
  }

  private func configure(
    _ source: [String: Any], key: String, variant: String, fit: String?, placeholder: String,
    attempt: Int, token: Int
  ) {
    guard token == self.epoch else { return }
    var completed = false
    let headers = source.dictionary("headers").flatMap {
      try? JSONSerialization.data(withJSONObject: $0, options: [.sortedKeys])
    }
    self.view.configure(
      sourceUri: source.string("uri").trimmingCharacters(in: .whitespacesAndNewlines),
      sourceHeadersJson: headers.flatMap { String(data: $0, encoding: .utf8) },
      variant: variant, contentFit: fit ?? source.string("contentFit", default: "cover"),
      cachePolicy: source.string("cachePolicy", default: "memory-disk"),
      autoplay: source.bool("autoplay"),
      recyclingKey: attempt == 0 ? key : "\(key):retry:\(attempt)",
      optimizeTos: attempt == 0 && source.bool("optimizeTos", default: true),
      overscan: source.double("overscan", default: 1.1),
      loadingStrategy: source.string("loadingStrategy", default: "none"),
      placeholderColor: placeholder,
      onLoad: { [weak self] in
        guard let self, self.epoch == token, !completed else { return }
        completed = true
        self.retry?.cancel()
        self.retry = nil
        self.result = true
        self.completion?(true)
      },
      onError: { [weak self] in
        guard let self, self.epoch == token, !completed else { return }
        completed = true
        self.result = false
        self.completion?(false)
        guard attempt < source.int("retryTimes"), self.retry == nil else { return }
        let work = DispatchWorkItem { [weak self] in
          guard let self, self.epoch == token else { return }
          self.retry = nil
          self.configure(
            source, key: key, variant: variant, fit: fit, placeholder: placeholder,
            attempt: attempt + 1, token: token)
        }
        self.retry = work
        DispatchQueue.main.asyncAfter(
          deadline: .now() + Double(Int.random(in: 0...2)), execute: work)
      })
  }

  func recycle() {
    epoch &+= 1
    retry?.cancel()
    retry = nil
    completion = nil
    identity = nil
    result = nil
    view.prepareForReuse()
  }
  deinit { retry?.cancel() }
}

enum NativeListSourceFallbackState {
  private static let sources: NSCache<NSString, NSNumber> = {
    let cache = NSCache<NSString, NSNumber>()
    cache.countLimit = 128
    return cache
  }()

  static func has(_ key: String) -> Bool {
    sources.object(forKey: key as NSString) != nil
  }

  static func remember(_ key: String) {
    sources.setObject(NSNumber(value: true), forKey: key as NSString)
  }

  static func forget(_ key: String) {
    sources.removeObject(forKey: key as NSString)
  }
}

func nativeListSourceFallbackStateKey(_ source: [String: Any]) -> String? {
  let uri = source.string("uri").trimmingCharacters(in: .whitespacesAndNewlines)
  guard !uri.isEmpty else { return nil }
  // The fallback-state cache is process-wide, so key it by a digest of the request identity
  // instead of retaining raw header values, which can carry credentials such as Authorization.
  let headers = source.dictionary("headers") ?? [:]
  let fields: [(name: String, value: String)] = headers.map { key, value in
    (name: key.lowercased(), value: String(describing: value))
  }
  let sortedFields = fields.sorted { lhs, rhs in
    lhs.name == rhs.name ? lhs.value < rhs.value : lhs.name < rhs.name
  }
  var canonicalValue = "\(uri.utf8.count):\(uri)"
  for field in sortedFields {
    canonicalValue +=
      "\(field.name.utf8.count):\(field.name)\(field.value.utf8.count):\(field.value)"
  }
  let digest = SHA256.hash(data: Data(canonicalValue.utf8))
  return digest.map { String(format: "%02x", $0) }.joined()
}

File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/NativeListImageSlot.kt
package com.margelo.nitro.nativelist

import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.onekeyimage.OneKeyImageReusableView
import org.json.JSONArray
import org.json.JSONObject

internal fun nativeListCanonical(value: Any?): String =
  when (value) {
    is JSONObject ->
      value.keys().asSequence().sorted().joinToString(",", "{", "}") {
        JSONObject.quote(it) + ":" + nativeListCanonical(value.opt(it))
      }
    is JSONArray ->
      (0 until value.length()).joinToString(",", "[", "]") { nativeListCanonical(value.opt(it)) }
    is String -> JSONObject.quote(value)
    else -> value?.toString() ?: "null"
  }

/** A request survives text-only binding; source/row/slot changes invalidate it. */
internal class NativeListImageSlot(context: ThemedReactContext) {
  val view = OneKeyImageReusableView(context)
  private var identity: String? = null
  private var epoch = 0
  private var retry: Runnable? = null
  private var completion: ((Boolean) -> Unit)? = null
  private var result: Boolean? = null

  fun bind(
    source: JSONObject,
    key: String,
    variant: String = "generic",
    fit: String? = null,
    placeholder: String = "#0000000F",
    round: Boolean = false,
    completion: ((Boolean) -> Unit)? = null,
  ) {
    this.completion = completion
    val next =
      nativeListCanonical(
        JSONObject()
          .put("source", source)
          .put("key", key)
          .put("variant", variant)
          .put("fit", fit ?: source.optString("contentFit", "cover"))
          .put("placeholder", placeholder)
          .put("round", round)
      )
    if (identity == next) {
      result?.let { completion?.invoke(it) }
      return
    }
    recycle()
    this.completion = completion
    identity = next
    configure(source, key, variant, fit, placeholder, round, 0, epoch)
  }

  private fun configure(
    source: JSONObject,
    key: String,
    variant: String,
    fit: String?,
    placeholder: String,
    round: Boolean,
    attempt: Int,
    token: Int,
  ) {
    if (epoch != token) return
    var completed = false
    view.configure(
      source.optString("uri").trim(),
      source.optJSONObject("headers")?.toString(),
      variant,
      fit ?: source.optString("contentFit", "cover"),
      source.optString("cachePolicy", "memory-disk"),
      source.optBoolean("autoplay"),
      if (attempt == 0) key else "$key:retry:$attempt",
      attempt == 0 && source.optBoolean("optimizeTos", true),
      source.optDouble("overscan", 1.1),
      source.optString("loadingStrategy", "none"),
      placeholder,
      round = round,
      onLoad = {
        if (epoch == token && !completed) {
          completed = true
          retry?.let(view::removeCallbacks)
          retry = null
          result = true
          completion?.invoke(true)
        }
      },
      onError = {
        if (epoch == token && !completed) {
          completed = true
          result = false
          completion?.invoke(false)
          if (attempt < source.optInt("retryTimes") && retry == null) {
            val task = Runnable {
              if (epoch == token) {
                retry = null
                configure(source, key, variant, fit, placeholder, round, attempt + 1, token)
              }
            }
            retry = task
            view.postDelayed(task, (0..2).random() * 1000L)
          }
        }
      },
    )
  }

  fun recycle() {
    epoch++
    retry?.let(view::removeCallbacks)
    retry = null
    identity = null
    completion = null
    result = null
    view.prepareForReuse()
  }

  fun dispose() {
    recycle()
    view.dispose()
  }
}

internal object NativeListSourceFallbackState {
  private const val CACHE_LIMIT = 128
  private val sources = LinkedHashMap<String, Unit>(CACHE_LIMIT, 0.75f, true)

  fun has(key: String): Boolean = synchronized(sources) { sources[key] != null }

  fun remember(key: String) {
    synchronized(sources) {
      sources[key] = Unit
      while (sources.size > CACHE_LIMIT) {
        sources.remove(sources.entries.first().key)
      }
    }
  }

  fun forget(key: String) {
    synchronized(sources) { sources.remove(key) }
  }
}
